src/System.Private.ServiceModel/src/System/ServiceModel/Channels/SingletonConnectionReader.cs

```diff
@@ -0,0 +1,612 @@
+            {
+                byte[] initialData = Fx.AllocateByteArray(_size);
+                Buffer.BlockCopy(buffer, _offset, initialData, 0, _size);
+                singletonConnection = new PreReadConnection(singletonConnection, initialData);
+            }
+
+            Stream connectionStream = new SingletonInputConnectionStream(this, singletonConnection, _transportSettings);
+            _inputStream = new MaxMessageSizeStream(connectionStream, _transportSettings.MaxReceivedMessageSize);
+            using (ServiceModelActivity activity = DiagnosticUtility.ShouldUseActivity ? ServiceModelActivity.CreateBoundedActivity(true) : null)
+            {
+                if (DiagnosticUtility.ShouldUseActivity)
+                {
+                    ServiceModelActivity.Start(activity, SR.Format(SR.ActivityProcessingMessage, TraceUtility.RetrieveMessageNumber()), ActivityType.ProcessMessage);
+                }
+
+                Message message = null;
+                try
+                {
+                    message = await _transportSettings.MessageEncoderFactory.Encoder.ReadMessageAsync(
+                        _inputStream, _transportSettings.MaxBufferSize, this.ContentType);
+                }
+                catch (XmlException xmlException)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(
+                        new ProtocolException(SR.Format(SR.MessageXmlProtocolError), xmlException));
+                }
+
+                if (DiagnosticUtility.ShouldUseActivity)
+                {
+                    TraceUtility.TransferFromTransport(message);
+                }
+
+                PrepareMessage(message);
+
+                return message;
+            }
+        }
+
+        // ensures that the reader is notified at end-of-stream, and takes care of the framing chunk headers
+        private class SingletonInputConnectionStream : ConnectionStream
+        {
+            private SingletonMessageDecoder _decoder;
+            private SingletonConnectionReader _reader;
+            private bool _atEof;
+            private byte[] _chunkBuffer; // used for when we have overflow
+            private int _chunkBufferOffset;
+            private int _chunkBufferSize;
+            private int _chunkBytesRemaining;
+
+            public SingletonInputConnectionStream(SingletonConnectionReader reader, IConnection connection,
+                IDefaultCommunicationTimeouts defaultTimeouts)
+                : base(connection, defaultTimeouts)
+            {
+                _reader = reader;
+                _decoder = new SingletonMessageDecoder(reader.StreamPosition);
+                _chunkBytesRemaining = 0;
+                _chunkBuffer = new byte[IntEncoder.MaxEncodedSize];
+            }
+
+            private void AbortReader()
+            {
+                _reader.Abort();
+            }
+
+            protected override void Dispose(bool disposing)
```

Comment:
Great - thanks for confirming

---
In reply to: [54129504](https://github.com/dotnet/wcf/pull/847#discussion_r54129504) [](ancestors = 54129504)